File: csharp/LINQtoObjects.Test/PaymentService/GetHighestPaidEmployeePerCompanyTests.cs
using LiveCodingExercises.LINQtoObjects.PaymentService;
using System;
using System.Collections.Generic;
using System.Linq;
using Xunit;

namespace LiveCodingExercises.LINQtoObjects.Test.PaymentService
{
    public class GetHighestPaidEmployeePerCompanyTests
    {
        private LINQtoObjects.PaymentService.PaymentService cut = new LINQtoObjects.PaymentService.PaymentService();

        [Fact]
        public void Should_return_data_for_all_companies_that_have_employees()
        {
            var result = cut.GetHighestPaidEmployeePerCompany(CreateCompanies());
            Assert.True(result.ContainsKey("A"));
            Assert.True(result.ContainsKey("B"));
            Assert.False(result.ContainsKey("C"));
        }

        [Fact]
        public void Should_return_highest_paid_employe_for_every_company_that_has_employees()
        {
            var result = cut.GetHighestPaidEmployeePerCompany(CreateCompanies());
            Assert.Contains(result["A"].Name, new List<string>(["David", "Eve"]));
            Assert.Equal("Lydia", result["B"].Name);
        }

        private IEnumerable<Company> CreateCompanies()
        {
            return [
                new Company("A", [
                    new Department("IT", [
                        new Employee("Alice", 5000m),
                        new Employee("Bob", 4000m),
                        new Employee("Charlie", 3000m),
                        new Employee("David", 7000m)
                        ]),
                    new Department("HR", [
                        new Employee("Eve", 7000m),
                        new Employee("Frank", 6000m),
                        new Employee("Georgia", 4500m),
                        new Employee("Henry", 2500m)
                        ])
                    ]),
                new Company("B", [
                    new Department("Sales", [
                        new Employee("Ike", 3000m),
                        new Employee("Janet", 4000m),
                        new Employee("Kate", 3500m),
                        new Employee("Lydia", 4500m)
                    ])
                ]),
                new Company("C", [
                    new Department("Cleaning", [])
                    ])
            ];
        }

    }
}

File: java/JavaStreams/src/test/java/pl/sebcel/livecoding/javastreams/paymentservice/GetHighestPaidEmployeePerCompanyTests.java
package pl.sebcel.livecoding.javastreams.paymentservice;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.Test;

public class GetHighestPaidEmployeePerCompanyTests {
	
	private PaymentService cut = new PaymentService();
	
	@Test
	public void should_throw_NullPointerException_when_list_of_companies_is_null() {
		assertThrows(NullPointerException.class, () -> {
			cut.getHighestPaidEmployeePerCompany(null);
		});
	}
	
	@Test
	public void should_throw_NullPointerException_when_list_of_departments_for_a_company_is_null() {
		Company companyWithNullDepartments = new Company("A", null);
		assertThrows(NullPointerException.class, () -> {
			cut.getHighestPaidEmployeePerCompany(List.of(companyWithNullDepartments));
		});
	}

	@Test
	public void should_throw_NullPointerException_when_list_of_employees_for_a_department_is_null() {
		Department departmentWithNullEmployees = new Department("Null", null);
		Company company = new Company("A", List.of(departmentWithNullEmployees));
		assertThrows(NullPointerException.class, () -> {
			cut.getHighestPaidEmployeePerCompany(List.of(company));
		});
	}

	@Test
	public void should_return_data_for_all_companies_that_have_employees() {
		Map<String, Employee> result = cut.getHighestPaidEmployeePerCompany(createCompanies());
		assertTrue(result.containsKey("A"));
		assertTrue(result.containsKey("B"));
		assertFalse(result.containsKey("C"));
	}
	
	@Test
	public void should_return_highest_paid_employee_per_each_company_that_has_employees() {
		Map<String, Employee> result = cut.getHighestPaidEmployeePerCompany(createCompanies());
		assertEquals("David", result.get("A").name());
		assertEquals("Lydia", result.get("B").name());
	}

	private List<Company> createCompanies() {
		return List.of(
				com("A",
						dep("HR",
								emp("Alice", "5000"),
								emp("Bob", "4000"),
								emp("Charlie", "3000"),
								emp("David", "7000")),
						dep("IT", 
								emp("Eve", "7000"),
								emp("Frank", "6000"),
								emp("Georgia", "4500"),
								emp("Henry", "2500"))
						),
				com("B",
						dep("Sales",
								emp("Ike","3000"),
								emp("Janet","4000"),
								emp("Kate","3500"),
								emp("Lydia","4500"))
						),
				com("C",
						dep("Cleaning"))
				);
	}
	
	private Employee emp(String name, String salary) {
		return new Employee(name, new BigDecimal(salary));
	}
	
	private Department dep(String name, Employee... employees) {
		return new Department(name, List.of(employees));
	}

	private Company com(String name, Department... departments) {
		return new Company(name, List.of(departments));
	}
}
